content/posts/2021-11-13---Typy.md

```diff
@@ -0,0 +1,292 @@
+}
+```
+Co dostaniemy na wyjściu tego programu? Odpowiedź jest oczywista! Obiekt test korzysta z pola JAVA z danego enuma. Nic więc dziwnego, że wykona się pierwszy if i dostaniemy wynik:
+```
+O tym właśnie jest ten BLOG :)
+
+Process finished with exit code 0
+```
+No bo właśnie o tym jest... :)
+W przeciwnym razie jak obiekt test przyjmie inną wartość, np PHP to if się nie wykona:
+```
+public class Main {
+    public static void main(String[] args) {
+        Languages.NaszEnum test = Languages.NaszEnum.PHP;
+        if(test.equals(Languages.NaszEnum.JAVA)){
+            System.out.println("O tym właśnie jest ten BLOG :)");
+        }
+        else{
+            System.out.println("Nie tego się aktualnie uczymy :(");
+        }
+    }
+}
+```
+a na wyjściu dostaniemy:
+```
+Nie tego się aktualnie uczymy :(
+
+Process finished with exit code 0
+```
+Bo PHP się właśnie nie uczmy. Całkiem proste prawda?
+
+### Recordy
+Ostatnim przystankiem tego bloga będą recordy. Stosunkowo świeży typ w javie, dodany dopiero w wersji 14.
+Tworząc klasę z przykładowo 4 polami, faktycznie wypisujemy tylko te 4 pola. Takie rzeczy jak konstruktory, gettery, settery czy np. metoda toString(),
+może być generowana automatycznie w naszym środowisku. Tutaj z pomocą przychodzą właśnie recordy. Ich deklaracja jest również nieco inna niż klasy, co pokazuję poniżej:
+```
+public record Route(int kilometers,
+                    String departure,
+                    String arrival,
+                    double fuelCost) {}
+```
+W tym przykładzie widzimy utworzony rekord Route z czterema polami: odległość w kilometrach, miejsce startu, miejsce zakończenia podróży oraz koszt paliwa.
+Parametry te znalazły się w nawiasach okrągłych, wypisane w oddzielnych wersach, lecz równie dobrze możemy to zapisać tak:
+```
+ public record Route(int kilometers, String departure, String arrival, double fuelCost) {}
+```
+Ten zapis również jest poprawny, ale tak jak zaznaczam po raz kolejny - przyjęło się to zapisywać troszkę inaczej podobnie jak np. w strumieniach.
+Rekordy charakteryzują się tym, że takie rzeczy jak konstruktor czy nadpisane metody jak np. toString() są już zawarte w takim rekordzie,
+dzięki czemu nasz kod jest bardziej przejrzysty dla oka. Zabrakło w tym przypadku getterów i setterów. Jest to spowodowane tym, że obiekty w rekordach są niemodyfikowalne,
+czyli nie możemy zmienić później ich wartości np. stosując settery. Do pól odwołujemy się później również troszkę inaczej, po prostu zamieszczając nazwę konkretnego pola po kropce przy tworzeniu obiektu:
+```
+ Route route = new Route(300, "Poznań", "Warszawa", 150.4);
+ double fuelCost = route.fuelCost();
+ System.out.println("Ten wyjazd kosztuję mnie "+fuelCost+" zł.");
+```
+Po skompilowaniu rekordu, zamienia się on w klasę finalną - przez co nie można po niej dziedziczyć.
+Rekordy umożliwiają nam również nadpisywanie implementacji metod takich jak hashCode(), czy toString() oraz definiowanie własnych klas staycznych.
```

Comment:
Recordy już mają hashCode, equals, toString dlatego je lubimy, więc warto o tym. Nie musimy tego nadpisywać, a wręcz mija się to z celem